dev-packages/e2e-tests/lib/getTestMatrix.ts

```
@@ -0,0 +1,151 @@
+}
+
+/**
+ * This methods generates a matrix for the GitHub Actions workflow to run the E2E tests.
+ * It checks which test applications are affected by the current changes in the PR and then generates a matrix
+ * including all test apps that have at least one dependency that was changed in the PR.
+ * If no `--base=xxx` is provided, it will output all test applications.
+ *
+ * If `--optional=true` is set, it will generate a matrix of optional test applications only.
+ * Otherwise, these will be skipped.
+ */
+function run(): void {
+  const args: Record<string, string> = {};
+  process.argv
+    .slice(2)
+    .filter(arg => arg.startsWith('--') && arg.includes('='))
+    .forEach(arg => {
+      const [part1, part2] = arg.split('=') as [string, string];
+      const argName = part1.replace('--', '');
+      const argValue = part2;
+      args[argName] = argValue;
+    });
+
+  const { base, head, optional = 'false' } = args;
+
+  const testApplications = globSync('*', { cwd: `${__dirname}/../test-applications/` });
+
+  // If `--base=xxx` is defined, we only want to get test applications changed since that base
+  // Else, we take all test applications (e.g. on push)
+  const includedTestApplications = base
+    ? getAffectedTestApplications(testApplications, { base, head })
+    : testApplications;
+
+  const optionalMode = optional === 'true';
+  const includes: MatrixInclude[] = [];
+
+  includedTestApplications.forEach(testApp => {
+    addIncludesForTestApp(testApp, includes, { optionalMode });
+  });
+
+  // We print this to the output, so the GHA can use it for the matrix
+  // eslint-disable-next-line no-console
+  console.log(`matrix=${JSON.stringify({ include: includes })}`);
+}
+
+function addIncludesForTestApp(
+  testApp: string,
+  includes: MatrixInclude[],
+  { optionalMode }: { optionalMode: boolean },
+): void {
+  const packageJson = getPackageJson(testApp);
+
+  // this means something went wrong
+  if (!packageJson) {
+    return;
```

l: We should throw an error in this case instead - fail fast.